src/compiler/transformer.ts

```diff
@@ -299,6 +310,31 @@ namespace ts {
             lexicalEnvironmentSuspended = false;
         }
 
+        function hasLeadingComments(node: Node) {
+            node = getParseTreeNode(node);
+            if (!node) {
+                return node && node.emitNode && !!node.emitNode.leadingComments;
+            }
+            const ranges = getLeadingCommentRangesOfNode(node, getSourceFileOfNode(node));
+            return ranges && ranges.length > 0;
+        }
+
+        function chunkVariableDeclarationsBySyntheticComments(declarations: VariableDeclaration[]): VariableDeclaration[][] {
+            return groupBy(declarations, decl => decl.emitNode && decl.emitNode.commentContext && hasLeadingComments(decl.emitNode.commentContext)
+                ? "" + getNodeId(decl.emitNode.commentContext)
+                : "no-comment"

```

Comment:
You could make `groupBy` use integer keys instead of strings and use a sparse array instead of a `Map`.